backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/report/DocOutboundConfigRepository.java

```diff
@@ -0,0 +1,169 @@
+import org.adempiere.model.InterfaceWrapperHelper;
+import org.adempiere.util.proxy.Cached;
+import org.compiere.model.I_AD_Column;
+import org.compiere.model.I_AD_PrintFormat;
+import org.compiere.model.I_C_Doc_Outbound_Config;
+import org.compiere.model.I_C_Doc_Outbound_Config_CC;
+import org.compiere.util.Env;
+import org.springframework.stereotype.Repository;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+
+import static de.metas.common.util.CoalesceUtil.coalesce;
+
+@Repository
+public class DocOutboundConfigRepository
+{
+	public static final IQueryBL queryBL = Services.get(IQueryBL.class);
+
+	public I_C_Doc_Outbound_Config getConfigById(final int docOutboundConfigId)
+	{
+		return InterfaceWrapperHelper.load(docOutboundConfigId, I_C_Doc_Outbound_Config.class);
+	}
+
+	// note that this method doesn't directly access the DB. Therefore, a unit test DAO implementation can extend this
+	// class without problems.
+	public final I_C_Doc_Outbound_Config retrieveConfig(final Properties ctx, final int tableId)
+	{
+		Check.assume(tableId > 0, "tableId > 0");
+
+		final int adClientId = Env.getAD_Client_ID(ctx);
+
+		I_C_Doc_Outbound_Config configSys = null;
+		I_C_Doc_Outbound_Config config = null;
+		for (final I_C_Doc_Outbound_Config currentConfig : retrieveAllConfigs())
+		{
+			if (currentConfig.getAD_Table_ID() == tableId)
+			{
+				if (currentConfig.getAD_Client_ID() == adClientId)
+				{
+					throwExceptionIfNotNull(config, tableId, adClientId, currentConfig);
+					config = currentConfig;
+				}
+				else if (currentConfig.getAD_Client_ID() == 0) // system
+				{
+					throwExceptionIfNotNull(configSys, tableId, adClientId, currentConfig);
+					configSys = currentConfig;
+				}
+			}
+		}
+		return coalesce(config, configSys);
+	}
+
+	private void throwExceptionIfNotNull(
+			@Nullable final I_C_Doc_Outbound_Config alreadyFoundConfig,
+			final int tableId,
+			final int adClientId,
+			@NonNull final I_C_Doc_Outbound_Config currentConfig)
+	{
+		if (alreadyFoundConfig == null)
+		{
+			return;
+		}
+		final String msg = StringUtils.formatMessage(
+				"Only one configuration shall exist for tableId '{}' on client '{}' but we found: {}, {}",
+				tableId, adClientId, alreadyFoundConfig, currentConfig);
+
+		throw new AdempiereException(msg)
+				.markAsUserValidationError(); // this error message is not exactly nice, but we still need to inform the user
+	}
+
+	public final I_C_Doc_Outbound_Config retrieveConfigForModel(@NonNull final Object model)
+	{
+		final Properties ctx = InterfaceWrapperHelper.getCtx(model);
+		final int adTableId = InterfaceWrapperHelper.getModelTableId(model);
+
+		return retrieveConfig(ctx, adTableId);
+	}
+
+	@Cached(cacheName = I_C_Doc_Outbound_Config_CC.Table_Name + "#AD_PrintFormatAll")
+	public List<PrintFormatId> retrieveAllPrintFormatIds(final int docOutboundConfigId)
```

Comment:
generally, for all these list-returning methods: better return `com.google.common.collect.ImmutableList`